File: src/1-creational/4-builder/no-builder.cs
public class Logger
{
    private Formatter _formatter;
    private Writer _writer;

    public void SetFormatter(Formatter formatter)
    {
        _formatter = formatter;
    }

    public void SetWriter(Writer writer)
    {
        if (_formatter == null)
        {
            throw new Exception("Need a formatter");
        }

        if (_formatter is JsonFormatter && writer is DatabaseWriter)
        {
            throw new Exception("Incompatible formatter for this writer");
        }

        _writer = writer;
    }

    public void Log(LogEntry entry)
    {
        if (_writer == null || _formatter == null)
        {
            throw new Exception("Logger is not configured");
        }

        _writer.Write(_formatter.Format(entry));
    }
}

class Application {
  public void Main() {
    var logger = new Logger();
    logger.SetWriter(new FileWriter()); // ! 😱 throws "Need a formatter"
    logger.SetFormatter(new JsonFormatter());
    logger.SetWriter(new DatabaseWriter()); // ! 😱 throws "Incompatible formatter for this writer"
    logger.Log(new { message = "Hello world!" });
    // ! 😱 you must remember to call the methods in the correct order,
    // ! and do it every time you need a new instance
  }
}

File: src/1-creational/4-builder/no-builder.ts
import { ConsoleWriter, DatabaseWriter, Formatter, JsonFormatter, LogEntry, Writer } from "./builder.dependencies";

// ! ❌ Bad example of not using a builder
class Logger {
  private formatter: Formatter | undefined;
  private writer: Writer | undefined;

  setFormatter(formatter: Formatter): void {
    this.formatter = formatter;
  }
  setWriter(writer: Writer): void {
    if (!this.formatter) {
      throw "Need a formatter";
    }
    if (this.formatter instanceof JsonFormatter && writer instanceof DatabaseWriter) {
      throw "Incompatible formatter for this writer";
    }
    this.writer = writer;
  }

  log(entry: LogEntry) {
    if (!this.writer || !this.formatter) {
      throw new Error("Logger is not configured");
    }
    this.writer.write(this.formatter.format(entry));
  }
}

class Application {
  main() {
    const logger = new Logger();
    // logger.setWriter(new FileWriter()); // ! 😱 throws "Need a formatter"
    logger.setFormatter(new JsonFormatter());
    // logger.setWriter(new DatabaseWriter()); // ! 😱 throws "Incompatible formatter for this writer"
    logger.setWriter(new ConsoleWriter());
    logger.log({ message: "Hello world!" });
    // ! 😱 you must remember to call the methods in the correct order,
    // ! and do it every time you need a new instance
  }
}

class Consumer {
  doWork() {
    const logger = new Logger();
    logger.setFormatter(new JsonFormatter());
    logger.log({ message: "Hello world!" });
  }
}
